TermGen.fs

```diff
@@ -18,193 +20,406 @@ open Starling.Core.Command
+                        \  (g2 ^ !(g1 ^ xbar=ybar)    -> v(ybar)^k)))   =qfail
+
+                   Our first task is to decide what each guarded func
+                   (rsucc, rfail, qsucc, qfail) is. *)
+
+                let barEq = List.map2 mkEq xbar ybar |> mkAnd
+
+                let rSuccG = mkAnd [ g1; g2; barEq; mkGt n k ]
+                let rSucc =
+                    { Func = { Cond = rSuccG; Item = func v xbar };
+                      Iterator = mkSub2 n k }
+                let rFailG = mkAnd [ g1; mkNot (mkAnd2 g2 barEq) ]
+                let rFail =
+                    { Func = { Cond = rFailG; Item = func v xbar }
+                      Iterator = n }
+                let qSuccG = mkAnd [ g2; g1; barEq; mkGt k n ]
+                let qSucc =
+                    { Func = { Cond = qSuccG; Item = func v ybar }
+                      Iterator = mkSub2 k n }
+                let qFailG = mkAnd [ g2; mkNot (mkAnd2 g1 barEq) ]
+                let qFail =
+                    { Func = { Cond = qFailG; Item = func v ybar }
+                      Iterator = k }
+
+                (* Now, we have to minus qSucc and qFail from rnext.  The first
+                   one is done in isolation (because eg. if we added rdone we'd
+                   accidentally calculate rdone/qFail!), but we can optimise the
+                   second one by passing through all the already-minused bits of
+                   (rdone, rSucc, rFail) as the new rdone. *)
+
+                // No need to check whether qSuccG/qFailG are trivially false:
+                // it's the first thing we'll do in each recursive call.
+                let innerMinus = minusViewByFunc qSucc rnext Multiset.empty
+
+                (* rSucc and rFail now get added to rdone for the tail call,
+                   but we can optimise here by not doing so if their guards are
+                   trivially false or their iterators are zero.
+
+                   Iterators can become negative here, but the n>k guards
+                   will always evaluate to false in this case, so they don't
+                   ever make it past here. *)
+                let optAdd rdoneSoFar rToAddG rToAdd =
+                    if isFalse rToAddG || rToAdd.Iterator = AInt 0L
+                    then rdoneSoFar
+                    else Multiset.add rdoneSoFar rToAdd
+
+                minusViewByFunc
+                    qFail
+                    innerMinus
+                    (optAdd (optAdd rdone rSuccG rSucc) rFailG rFail)
+
+/// <summary>
+///     Generates the frame part of the weakest precondition.
+///     <para>
+///         In the meta-theory, this is <c>R \ Q</c>.
+///     </para>
+/// </summary>
+/// <param name="r">
+///     The view representing the goal to be subtracted from.
+/// </param>
+/// <param name="q">
+///     The view representing the postcondition to subtract.
+/// </param>
+/// <returns>
+///     The subtracted frame view.
+/// </returns>
 let termGenFrame
```

Comment:
Changed it to `termGenWPreMinus` for now, and the function below to `termGenWPre`.  Still unsure.